feat(Analysis): add improper integral of complex exp

wwylele opened the pull request:
These are the infinite version of integral_exp_mul_complex.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/ImproperIntegrals.lean:
```diff
@@ -35,6 +35,13 @@ theorem integrableOn_exp_Iic (c : ℝ) : IntegrableOn exp (Iic c) := by
   simp_rw [norm_of_nonneg (exp_pos _).le, integral_exp, sub_le_self_iff]
   exact (exp_pos _).le
 
+theorem integrableOn_exp_neg_Ioi (c : ℝ) : IntegrableOn (fun (x : ℝ) => exp (-x)) (Ioi c) := by
```
I think you should prove an analogue of [MeasureTheory.Integrable.comp_neg](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Group/Integral.html#MeasureTheory.Integrable.comp_neg) for `IntegrableOn` and general sets, and likewise for everything else in that file. Then just use it here.

And why not `Ici c`?

wwylele replied:
Yeah I'll try extracting a `IntegrableOn.comp_neg`-like theorem. As for why not `Ici`, it seems that most theorems are only stated for `Ioi` and `Iic`, like existing `integral_exp...` in this file. But if we think it is better to have all 4 versions, I certainly can add them

j-loreaux replied:
Sorry, I meant to take out that "Why not ...?" but forgot. Please note my other comment about adding an explanation for why we don't include the others.

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/ImproperIntegrals.lean:
```diff
@@ -51,6 +58,54 @@ theorem integral_exp_neg_Ioi (c : ℝ) : (∫ x : ℝ in Ioi c, exp (-x)) = exp
+    apply (integrableOn_Ioi_comp_mul_left_iff (fun x => exp (-x)) c a_neg).mpr
+    apply integrableOn_exp_neg_Ioi
+
+theorem integrableOn_exp_mul_complex_Iic {a : ℂ} (ha : 0 < a.re) (c : ℝ) :
+    IntegrableOn (fun x : ℝ => Complex.exp (a * x)) (Iic c) := by
```
This should follow from the above lemma and the `integrableOn.comp_neg` lemma I mentioned.

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/ImproperIntegrals.lean:
```diff
@@ -51,6 +58,54 @@ theorem integral_exp_neg_Ioi (c : ℝ) : (∫ x : ℝ in Ioi c, exp (-x)) = exp
 theorem integral_exp_neg_Ioi_zero : (∫ x : ℝ in Ioi 0, exp (-x)) = 1 := by
   simpa only [neg_zero, exp_zero] using integral_exp_neg_Ioi 0
 
+theorem integrableOn_exp_mul_complex_Ioi {a : ℂ} (ha : a.re < 0) (c : ℝ) :
+    IntegrableOn (fun x : ℝ => Complex.exp (a * x)) (Ioi c) := by
```
You should have remark like the one [here](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Lebesgue/Integral.html#Substituting-x-for-x) describing why you only have `Ioi` and `Iic` (although, for the `IntegrableOn` statements, as opposed to the integrals themselves, I suspect it's better to have closed infinite intervals everywhere, but I may be wrong.)

j-loreaux reviewed (commented):
It looks like you're working harder than necessary in the proofs of the value of the integrals, but I'm out of time for review now so I'm not sure.

wwylele commented:
I have addressed the comments:
 - I observed that the entire file is following the convention to only state lemmas for `Iic` and `Ioi`, so I copied the note from https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Lebesgue/Integral.html#Substituting-x-for-x
 - I extracted `MeasureTheory.IntegrableOn.comp_neg_Iic` and `MeasureTheory.IntegrableOn.comp_neg_Ioi` and put them together with [`integral_comp_neg_Iic`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Lebesgue/Integral.html#integral_comp_neg_Iic). These are specifically for real domain instead of more general sets/groups. I didn't find a nice way to extract a lemma for general groups, as it becomes a thin specialization of `integrable_map_equiv` with `MeasurableEquiv.neg` and it feels more cumbersome to use than just using `integrable_map_equiv` itself...

j-loreaux commented on Mathlib/MeasureTheory/Measure/Lebesgue/Integral.lean:
```diff
@@ -74,6 +74,23 @@ mathlib's conventions for integrals over finite intervals (see `intervalIntegral
 of finite integrals, see `intervalIntegral.integral_comp_neg`.
 -/
 
+theorem MeasureTheory.IntegrableOn.comp_neg_Iic {E : Type*} [NormedAddCommGroup E]
+    {c : ℝ} {f : ℝ → E} (hf : IntegrableOn f (Ioi (-c))) :
+    IntegrableOn (fun x => f (-x)) (Iic c) := by
```
Please prove the more general version which is analogous to: [MeasureTheory.Integrable.comp_neg](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Group/Integral.html#MeasureTheory.Integrable.comp_neg). It should be `(hf : IntegrableOn f s) : IntegrableOn (fun x ↦ f (-x)) -s)` (and you'll probably need `open scoped Pointwise` for the `-s` to make sense).

You can then prove both these lemmas as special cases of that one.

wwylele replied:
Thanks for the pointer. I think I get it now.

j-loreaux commented on Mathlib/MeasureTheory/Measure/Lebesgue/Integral.lean:
```diff
@@ -74,6 +75,17 @@ mathlib's conventions for integrals over finite intervals (see `intervalIntegral
+theorem MeasureTheory.IntegrableOn.comp_neg_Iic {E : Type*} [NormedAddCommGroup E]
+    {c : ℝ} {f : ℝ → E} (hf : IntegrableOn f (Ioi (-c))) :
+    IntegrableOn (fun x => f (-x)) (Iic c) := by
+  rw [(by simp : Iic c = - (Ici (-c)))]
+  exact (integrableOn_Ici_iff_integrableOn_Ioi.mpr hf).comp_neg
```
```suggestion
  simpa using (integrableOn_Ici_iff_integrableOn_Ioi.mpr hf).comp_neg
```

j-loreaux commented on Mathlib/MeasureTheory/Measure/Lebesgue/Integral.lean:
```diff
@@ -74,6 +75,17 @@ mathlib's conventions for integrals over finite intervals (see `intervalIntegral
+theorem MeasureTheory.IntegrableOn.comp_neg_Ioi {E : Type*} [NormedAddCommGroup E]
+    {c : ℝ} {f : ℝ → E} (hf : IntegrableOn f (Iic (-c))) :
+    IntegrableOn (fun x => f (-x)) (Ioi c) := by
+  rw [(by simp : Ioi c = - (Iio (-c)))]
+  exact (integrableOn_Iic_iff_integrableOn_Iio.mp hf).comp_neg
```
```suggestion
  simpa using (integrableOn_Iic_iff_integrableOn_Iio.mp hf).comp_neg
```

j-loreaux commented on Mathlib/MeasureTheory/Group/Integral.lean:
```diff
@@ -38,6 +38,14 @@ theorem integral_inv_eq_self (f : G → E) (μ : Measure G) [IsInvInvariant μ]
+theorem IntegrableOn.comp_inv [IsInvInvariant μ] {f : G → F} {s : Set G} (hf : IntegrableOn f s μ) :
+    IntegrableOn (fun x => f x⁻¹) s⁻¹ μ := by
+  apply (integrable_map_equiv (MeasurableEquiv.inv G) f).mp
+  rw [(by simp : s⁻¹ = MeasurableEquiv.inv G ⁻¹' s)]
+  rw [← MeasurableEquiv.restrict_map]
```
not necessary, but I would say this is slightly more idiomatic.
```suggestion
  have : s⁻¹ = MeasurableEquiv.inv G ⁻¹' s := by simp
  rw [this, ← MeasurableEquiv.restrict_map]
```

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/ImproperIntegrals.lean:
```diff
@@ -51,6 +57,49 @@ theorem integral_exp_neg_Ioi (c : ℝ) : (∫ x : ℝ in Ioi c, exp (-x)) = exp
+
+theorem integrableOn_exp_mul_complex_Iic {a : ℂ} (ha : 0 < a.re) (c : ℝ) :
+    IntegrableOn (fun x : ℝ => Complex.exp (a * x)) (Iic c) := by
+  conv in (fun x => _) =>
+    ext x
+    rw [(by simp : a * x = -a * (-x : ℝ))]
+  have a_neg : (-a).re < 0 := by simp [ha]
+  exact (integrableOn_exp_mul_complex_Ioi a_neg (-c)).comp_neg_Iic
```
```suggestion
  simpa using (integrableOn_exp_mul_complex_Ioi (a := -a) (by simpa) (-c)).comp_neg_Iic
```

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/ImproperIntegrals.lean:
```diff
@@ -51,6 +57,49 @@ theorem integral_exp_neg_Ioi (c : ℝ) : (∫ x : ℝ in Ioi c, exp (-x)) = exp
+  refine (integrable_norm_iff ?_).mp ?_
+  · apply Continuous.aestronglyMeasurable
+    fun_prop
+  · conv in (fun x => _) =>
+      ext x
+      rw [(by simp [Complex.norm_exp] : ‖Complex.exp (a * x)‖ = exp (-(-a.re * x)))]
+    have a_neg : 0 < -a.re := by simp [ha]
+    apply (integrableOn_Ioi_comp_mul_left_iff (fun x => exp (-x)) c a_neg).mpr
+    apply integrableOn_exp_neg_Ioi
```
```suggestion
  · simpa [Complex.norm_exp] using
      (integrableOn_Ioi_comp_mul_left_iff (fun x => exp (-x)) c (a := -a.re) (by simpa)).mpr <|
        integrableOn_exp_neg_Ioi _ 
```

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/ImproperIntegrals.lean:
```diff
@@ -51,6 +57,49 @@ theorem integral_exp_neg_Ioi (c : ℝ) : (∫ x : ℝ in Ioi c, exp (-x)) = exp
+
+theorem integral_exp_mul_complex_Ioi {a : ℂ} (ha : a.re < 0) (c : ℝ) :
+    ∫ x : ℝ in Set.Ioi c, Complex.exp (a * x) = - Complex.exp (a * c) / a := by
+  refine
+    tendsto_nhds_unique (intervalIntegral_tendsto_integral_Ioi c
+      (integrableOn_exp_mul_complex_Ioi ha c) tendsto_id) ?_
+  have nonzero : a ≠ 0 := fun h => (Complex.zero_re ▸ h ▸ ha).false
+  simp_rw [integral_exp_mul_complex nonzero, id_eq]
+  rw [(by simp : - Complex.exp (a * c) / a = (0 - Complex.exp (a * c)) / a)]
+  refine Tendsto.div_const (Tendsto.sub_const ?_ _) _
+  apply Complex.tendsto_exp_nhds_zero_iff.mpr
+  simp only [Complex.mul_re, Complex.ofReal_re, Complex.ofReal_im, mul_zero, sub_zero]
+  exact Tendsto.neg_mul_atTop ha tendsto_const_nhds tendsto_id
```
1. `aesop` can prove the `have nonzero`, so we can inline it into the `simp_rw`.
2. Your `rw [(by simp...)]` should have just be `rw [← sub_zero]`, but...
3. sometimes it's nicer to just tell Lean the thing you want to prove with `suffices`
```suggestion
  refine tendsto_nhds_unique (intervalIntegral_tendsto_integral_Ioi c
    (integrableOn_exp_mul_complex_Ioi ha c) tendsto_id) ?_
  simp_rw [integral_exp_mul_complex (c := a) (by aesop), id_eq]
  suffices Tendsto (fun x : ℝ ↦ Complex.exp (a * x)) atTop (𝓝 0) by
    simpa using this.sub_const _ |>.div_const _
  simpa [Complex.tendsto_exp_nhds_zero_iff] using tendsto_const_nhds.neg_mul_atTop ha tendsto_id
```

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/ImproperIntegrals.lean:
```diff
@@ -51,6 +57,49 @@ theorem integral_exp_neg_Ioi (c : ℝ) : (∫ x : ℝ in Ioi c, exp (-x)) = exp
+
+theorem integral_exp_mul_complex_Iic {a : ℂ} (ha : 0 < a.re) (c : ℝ) :
+    ∫ x : ℝ in Set.Iic c, Complex.exp (a * x) = Complex.exp (a * c) / a := by
+  conv in (fun x => _) =>
+    ext x
+    rw [(by simp : a * x = -a * (-x : ℝ))]
+  rw [integral_comp_neg_Iic c (fun x => Complex.exp (-a * x))]
+  have a_neg : (-a).re < 0 := by simp [ha]
+  rw [integral_exp_mul_complex_Ioi a_neg]
+  simp
```
```suggestion
  simpa [neg_mul, ← mul_neg, ← Complex.ofReal_neg,
    integral_comp_neg_Ioi (f := fun x : ℝ ↦ Complex.exp (a * x))]
      using integral_exp_mul_complex_Ioi (a := -a) (by simpa) (-c)
```

j-loreaux reviewed (commented):
The suggestions below are not actually important, per se, they are primarily informative. It seems there are two things you can learn from these:

1. `simpa` is a great tactic. It's useful especially wheny you can build a slightly more complicated expression, and then use `simp` to rewrite it into the form you want.
2. You can provide an explicit value for an implicit argument `a` via the syntax `(a := my_value)`. This is useful, among other reasons, when you want to prove the explicit argument from which `a` is normally inferred using some automation (like `simpa` or `aesop`).

wwylele commented:
@j-loreaux Thanks for the review! I am definitely still on my way learning the idiomatic way in mathlib, but I have learned a lot

j-loreaux commented on Mathlib/MeasureTheory/Measure/Lebesgue/Integral.lean:
```diff
@@ -74,6 +75,15 @@ mathlib's conventions for integrals over finite intervals (see `intervalIntegral
 of finite integrals, see `intervalIntegral.integral_comp_neg`.
 -/
 
+theorem MeasureTheory.IntegrableOn.comp_neg_Iic {E : Type*} [NormedAddCommGroup E]
```
Presumably this and the following result generalize to the setting where the domain is a linearly ordered additive group, yes? Please generalize to that situation.

wwylele commented:
I generalized the `IntegrableOn.comp_inv_I**` theorems to ordered group. Because I want to avoid adding `NoAtoms`, I provided the full family of `Iic`, `Ici`, `Iio`, and `Ioi` there.

j-loreaux reviewed (commented):
Thanks! 

bors merge
